### PR TITLE
Support collect aggregations in reduction

### LINKED ISSUE #7807: [FEA] support collect aggregations in reduction
**Is your feature request related to a problem? Please describe.**
Currently, cuDF supports collect aggregations in rolling windows and groupBy context (though the support is not complete). But, collect aggregations in reduction context is still missing.

**Additional context**
I believe we can not support this feature until #5887 got solved, since scalar of ListType is essential in reduction to collect.

### PR DESCRIPTION
**Pull request by sperlingxx**
Closes #7807 

Curreent PR is to support the collect aggregation family in reduction context, which includes collect_list, collect_set, merge_lists, and merge_sets.
The implementations are inspired by corresponding collect aggregations in groupby context.

### CODE REVIEW

File: cpp/src/reductions/collect_ops.cu [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+std::unique_ptr<scalar> collect_list(column_view const& col,
+                                     null_policy null_handling,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  if (null_handling == null_policy::EXCLUDE && col.has_nulls()) {
+    auto not_null_pred = [mask = col.null_mask(), offset = col.offset()] __device__(auto i) {
+      return bit_is_set(mask, offset + i);
+    };
```
Comment (jrhemstad) [inline cpp/src/reductions/collect_ops.cu:50]: Use `make_validity_iterator`.

File: cpp/src/reductions/collect_ops.cu [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    auto null_purged_table  = cudf::detail::copy_if(table_view{{col}}, not_null_pred, stream, mr);
+    column* null_purged_col = null_purged_table->release().front().release();
+    null_purged_col->set_null_mask(rmm::device_buffer{0, stream, mr}, 0);
+    return std::make_unique<list_scalar>(*null_purged_col, true, stream, mr);
```
Comment (jrhemstad) [inline cpp/src/reductions/collect_ops.cu:54]: This makes a copy of `null_purged_col` and then leaks it. 
```suggestion
    auto null_purged_col = null_purged_table->release().front();
    null_purged_col->set_null_mask(rmm::device_buffer{0, stream, mr}, 0);
    return std::make_unique<list_scalar>(std::move(*null_purged_col), true, stream, mr);
```

File: cpp/src/reductions/collect_ops.cu (hunk 1 above)
Comment (sperlingxx) [reply cpp/src/reductions/collect_ops.cu]: Changed.

File: cpp/src/reductions/collect_ops.cu (hunk 2 above)
Comment (sperlingxx) [reply cpp/src/reductions/collect_ops.cu]: @jrhemstad  Change done. Thank you for the catch!

File: cpp/src/reductions/collect_ops.cu [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+std::unique_ptr<scalar> merge_sets(column_view const& col,
+                                   null_equality nulls_equal,
+                                   nan_equality nans_equal,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(col.type().id() == type_id::LIST,
+               "input column of merge_lists must be a list column");
```
Comment (jrhemstad) [inline cpp/src/reductions/collect_ops.cu:89]: It's more descriptive to just make the parameter a `lists_column_view` and eliminates the extra error checking. 
```suggestion
std::unique_ptr<scalar> merge_sets(lists_column_view const& col,
                                   null_equality nulls_equal,
                                   nan_equality nans_equal,
                                   rmm::cuda_stream_view stream,
                                   rmm::mr::device_memory_resource* mr)
{
```

File: cpp/src/reductions/collect_ops.cu [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    return make_list_scalar(col, stream, mr);
+  }
+}
+
+std::unique_ptr<scalar> merge_lists(column_view const& col,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(col.type().id() == type_id::LIST,
+               "input column of merge_lists must be a list column");
```
Comment (jrhemstad) [inline cpp/src/reductions/collect_ops.cu:65]:
```suggestion
std::unique_ptr<scalar> merge_lists(lists_column_view const& col,
                                    rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
{
```

File: cpp/src/reductions/collect_ops.cu (hunk 3 above)
Comment (sperlingxx) [reply cpp/src/reductions/collect_ops.cu]: Done.

**Review by jrhemstad — APPROVED** [review]

**Comment by sperlingxx** [comment]
@gpucibot merge

**Pull request merged**